### PR TITLE
add ms.author to errors.log

### PR DESCRIPTION
**Pull request by irushx**
###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6812)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/docfx?pullRequest=6812) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/docfx?pullRequest=6812) before we can accept your contribution.<br/><br/>:x: irushx [sign now](https://cla.dotnetfoundation.org/dotnet/docfx?pullRequest=6812)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/docfx?pullRequest=6812) it.</sub>

File: docs/specs/metadata.yml [hunk 1]
```diff
@@ -43,6 +43,19 @@ outputs:
   a.json: |
     { "ms.custom-metadata": null }
 ---
+# Add ms.author
```
Comment (yufeih) [inline docs/specs/metadata.yml:46]:
```suggestion
# Add ms.author to error log
```

File: src/docfx/build/metadata/UserMetadata.cs [hunk 2]
```diff
@@ -21,6 +21,9 @@ internal class UserMetadata
 
         public SourceInfo<string> Author { get; private set; } = new SourceInfo<string>("");
 
+        [JsonProperty("ms.author")]
+        public SourceInfo<string> MsAuthor { get; private set; } = new SourceInfo<string>("");
```
Comment (yufeih) [inline src/docfx/build/metadata/UserMetadata.cs:25]: make `ms.author` nullable so if a file does not have `ms.author`, it should not appear in `.errors.log`. Update test case to assert `ms.author` does not appear using `"ms.author": undefined"`
```suggestion
        public SourceInfo<string?> MsAuthor { get; private set; } = new SourceInfo<string?>();
```

File: src/docfx/lib/log/ErrorLog.cs [hunk 3]
```diff
@@ -30,7 +30,13 @@ public override void Add(Error error)
         {
             if (CustomRuleProvider != null)
             {
-                error = CustomRuleProvider.ApplyCustomRule(error);
+                try
+                {
+                    error = CustomRuleProvider.ApplyCustomRule(error);
+                }
+                catch (Exception ex) when (DocfxException.IsDocfxException(ex, out var dex))
+                {
```
Comment (yufeih) [inline src/docfx/lib/log/ErrorLog.cs:38]: Log the exception to console using `Log.Write(ex)`

File: src/docfx/validation/CustomRuleProvider.cs [hunk 4]
```diff
@@ -24,11 +26,13 @@ internal class CustomRuleProvider
             DocumentProvider documentProvider,
             Lazy<PublishUrlMap> publishUrlMap,
             MonikerProvider monikerProvider,
+            MetadataProvider metaDataProvider,
```
Comment (yufeih) [inline src/docfx/validation/CustomRuleProvider.cs:29]:
```suggestion
            MetadataProvider metadataProvider,
```

File: src/docfx/validation/CustomRuleProvider.cs [hunk 5]
```diff
@@ -64,6 +68,9 @@ public Error ApplyCustomRule(Error error)
             {
                 error = WithCustomRule(error, customRule);
             }
+            var source = error.Source?.File;
+            var userMeta = source != null ? _metadataProvider.GetMetadata(ErrorBuilder.Null, source) : null;
```
Comment (yufeih) [inline src/docfx/validation/CustomRuleProvider.cs:72]:
```suggestion
            var userMetadata = source != null ? _metadataProvider.GetMetadata(ErrorBuilder.Null, source) : null;
```

File: src/docfx/lib/log/Error.cs [hunk 6]
```diff
@@ -84,21 +96,21 @@ public override string ToString()
...
+                File = file,
+                Line = line,
+                EndLine = end_line,
+                Column = column,
+                EndColumn = end_column,
+                LogItemType = "user",
+                PullRequestOnly = PullRequestOnly ? (bool?)true : null,
+                PropertyPath = PropertyPath,
+                DateTime = DateTime.UtcNow, // Leave data_time as the last field to make regression test stable
+                MsAuthor = MsAuthor,
```
Comment (yufeih) [inline src/docfx/lib/log/Error.cs:113]: See comment in line 112, move `MsAuthor` before `DateTime`.

File: src/docfx/lib/log/Error.cs [hunk 7]
```diff
@@ -84,21 +96,21 @@ public override string ToString()
             var end_line = source?.EndLine ?? 0;
             var column = source?.Column ?? 0;
             var end_column = source?.EndColumn ?? 0;
-
-            return JsonUtility.Serialize(new
+            return JsonUtility.Serialize(new ErrorItem
```
Comment (yufeih) [inline src/docfx/lib/log/Error.cs:99]: nit: instead of creating a class, how about
```suggestion
            return JsonUtility.Serialize(new JObject
```

Comment (herohua) [reply src/docfx/lib/log/Error.cs:99]: Is it possible to support `[JsonProperty("ms.author")]` within a JObject?

---
In reply to: [530181798](https://github.com/dotnet/docfx/pull/6812#discussion_r530181798) [](ancestors = 530181798)

File: src/docfx/lib/log/Error.cs [hunk 8]
```diff
@@ -3,6 +3,10 @@
 
 using System;
 using System.Collections.Generic;
+using System.Text.Json;
+using Microsoft.DocAsCode.Common;
```
Comment (yufeih) [inline src/docfx/lib/log/Error.cs:7]: What is `Microsoft.DocAsCode.Common` for?

**Review by yufeih — APPROVED** [review]

**Review by yufeih — APPROVED** [review]
Love this dirty hack 😃

**Review by live1206 — APPROVED** [review]

**Comment by yufeih** [comment]
dryrun of _azure-docs-pr_ seems to start slower from https://github.com/dotnet/docfx/pull/6806/files

**Pull request merged**